### PR TITLE
Introduce GNTB

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: contracts/GolemNetworkToken2.sol [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2018 Golem Factory
+// Licensed under the GNU General Public License v3. See the LICENSE file.
+
+pragma solidity ^0.4.19;
+
+import "./ExtendedToken.sol";
```
Comment (Krigpl) [inline contracts/GolemNetworkToken2.sol]: I dunno if ExtendedToken is worth extracting after all, we could just move its methods into this new Token.

File: contracts/GolemNetworkToken2.sol [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import "./TokenProxy.sol";
+
+contract GolemNetworkToken2 is TokenProxy, ExtendedToken {
+
+    string public constant name = "Golem Network Token 2";
+    string public constant symbol = "GNT2";
+    uint8 public constant decimals = 18;
+
+    function GolemNetworkToken2(ERC20Basic _token) TokenProxy(_token) public {
```
Comment (Krigpl) [inline contracts/GolemNetworkToken2.sol]: `s/_token/_gntToken/` ?
This is no longer generic.

File: contracts/GolemNetworkToken2.sol (hunk 1 above)
Comment (chfast) [reply contracts/GolemNetworkToken2.sol]: I kinda agree, but might be easier to review this if generic implementation (like batchTransfer()) are separated.

File: contracts/GolemNetworkToken2.sol [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+// Licensed under the GNU General Public License v3. See the LICENSE file.
+
+pragma solidity ^0.4.19;
+
+import "./ExtendedToken.sol";
+import "./TokenProxy.sol";
+
+contract GolemNetworkToken2 is TokenProxy, ExtendedToken {
+
+    string public constant name = "Golem Network Token 2";
```
Comment (badb) [inline contracts/GolemNetworkToken2.sol]: Please don't. Let's keep the previous name or use Golem Network Token Batching, but not this. It's misleading.

File: contracts/GolemNetworkToken2.sol [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    string public constant symbol = "GNT2";
```
Comment (badb) [inline contracts/GolemNetworkToken2.sol]: Same here. GNTB, GWBT, whatever. We have alpha2 and golem 2.0 - that's enough.

File: contracts/GolemNetworkToken2.sol (hunk 1 above)
Comment (Krigpl) [reply contracts/GolemNetworkToken2.sol]: We'd also remove diamond inheritance which might be problematic. I honestly don't know how it works in Solidity.

File: contracts/GolemNetworkToken2.sol (hunk 3 above)
Comment (lukasz-glen) [reply contracts/GolemNetworkToken2.sol]: :)
GolemNetworkTokenNew

File: contracts/GolemNetworkToken2.sol (hunk 4 above)
Comment (chfast) [reply contracts/GolemNetworkToken2.sol]: Maybe Golem Network Token Whatever then?

File: contracts/GolemNetworkToken2.sol (hunk 1 above)
Comment (chfast) [reply contracts/GolemNetworkToken2.sol]: We will not. It would be `GNT2 is TokenProxy, StandardToken`.

File: contracts/GNTDeposit.sol [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
+import "./GolemNetworkToken2.sol";
 import "./ReceivingContract.sol";
 
 contract GNTDeposit is ReceivingContract {
     address public oracle;
     address public coldwallet;
     uint256 public withdrawal_delay;
 
-    GolemNetworkTokenWrapped public token;
+    GolemNetworkToken2 public token;
```
Comment (lukasz-glen) [inline contracts/GNTDeposit.sol]: Please consider this: to weaken type of token to ERC20 for instance.

File: contracts/GolemNetworkToken2.sol (hunk 1 above)
Comment (lukasz-glen) [reply contracts/GolemNetworkToken2.sol]: contract TokenProxy is ERC20Basic
would it be ok?

Comment (chfast) [reply contracts/GolemNetworkToken2.sol]: I will do it. Now or in the next PR?

File: contracts/GolemNetworkTokenBatching.sol [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+// Licensed under the GNU General Public License v3. See the LICENSE file.
+
+pragma solidity ^0.4.19;
+
+import "./ExtendedToken.sol";
+import "./TokenProxy.sol";
+
+contract GolemNetworkTokenBatching is TokenProxy, ExtendedToken {
+
+    string public constant name = "Golem Network Token Batching";
```
Comment (Krigpl) [inline contracts/GolemNetworkTokenBatching.sol]: I dunno we should focus on `batching` in the name, this token provides full ERC20 compliance and ERC677-like functionality as well expect for the `batchTransfer`.
I'm fine with the code and we can agree on the name later if needed.

File: contracts/GNTDeposit.sol [hunk 7]
```diff
@@ -1,14 +1,14 @@
 pragma solidity ^0.4.16;
 
-import "./GolemNetworkTokenWrapped.sol";
+import "./GolemNetworkTokenBatching.sol";
```
Comment (Krigpl) [inline contracts/GNTDeposit.sol]: `ERC20`

**Comment by chfast** [comment]
How about we merge it with the current name and you change it to different one later on?

I will also send another PR that simplifies contract hierarchy.

**Pull request merged**